extra_metadata takes user input for 2D diags of 3D vars

timothyas opened the pull request:
Title says it all! The only current problem is when 2D and 3D versions of the same variable exist in different files which are read in at the same time. 

See #142 for a list of new tests which need to be made. I would need some guidance in doing that, but I'd be happy to implement them.

timothyas commented:
Ok latest commit fixes the issue I mentioned. Now, if 2D diagnostics are recognized then the variable name gets appended with '_2D' within the dataset.

timothyas commented:
I actually think we could do this without the user input, but it requires an assumption. These lines are the main change 
https://github.com/xgcm/xmitgcm/blob/6dd33bdfe9efdb28a9915c09afc97a97833b9402/xmitgcm/mds_store.py#L686-L697

I noticed that now only lines 690 on are needed. They assume that 3 dims are expected, but only find two (or in the case of LLC which has face dim, it's 4 and 3), then collapse. I'm curious what you think - maybe this assumes too much and so should we still force the user to tell xmitgcm that they expect 2D slices since this is missing from the meta information? 

Thanks guys!

raphaeldussin commented:
I think this is great! Should we get started with the new tests?

timothyas commented:
Thanks @raphaeldussin. I'm curious if you think we still need the user to provide the flag '2d_diag_of_3d_var' even though I've found all the tests pass with just the two if statements (without the two involving `extra_metadata`): 

```
         #Expect 3 dims, vertical and 2 horizontal 
         ndims_expected = 3 
         if self.llc: 
             # llc geometry adds face dim 
             ndims_expected += 1 
  
         if len(dims) == 3 and data.ndim == ndims_expected-1: 
             dims = dims[1:] 
             vname = '%s_2D' % vname 
``` 

But like I said I'm worried this would be assuming too much - maybe I'm overthinking things! :P  

As for the tests, I have a rookie question... For reference here is what @rabernat originally suggested: 

- Copy one of the 2D files (like Eta) to a 3D file name (like Theta) and create the appropriate meta file. we should test both built-in state variables and those generated from available_diagnostics.log
- Open those files with open_mdsdataset and ensure no errors are raised
- Check that the files have the right dimensions.

For step 1, do you want to do this all within a test function? So that each time the tests run, it does the steps of copying Eta, making the meta file, writes it out to be loaded in. Or manually copy Eta etc and load it to wherever you guys keep your test data, so that it gets loaded like the other variables each time? If the latter, I'm curious where to load these meta/data files.  Thanks!

rabernat commented:
I think it would be useful to discuss the optimal design here in general terms, before working more on the implementation. **What do we want to happen?**

Here is the scenario. We have a file, let's say `Theta.data/meta` which, instead of having the expected 3 dimensions, has only 2 dimensions. This can happen if MITgcm diagnostics are configure to output only one k level.

_To preserve backwards compatibility, I would like an error to continue to be raised, as it is now._

However, I would like to add an optional flag like `allow_2d_diags=True` to `open_mdsdataset` which turns on the extra logic @timothyas provided above. After thinking about it more, I agree that `extra_metadata` doesn't have to be involved. 

I can try to tackle the tests. I'll push some proposed tests to your branch later today.

rabernat commented:
I just added this test:

https://github.com/xgcm/xmitgcm/blob/2fd7ad7a8f8c1077e314aa7a0ee54f1d9dbe6e37/xmitgcm/test/test_mds_store.py#L199-L234

This is not totally complete, but I think it should help @timothyas move forward.
The strange thing is that it passes just fine already with standard grids but doesn't work with LLC grids. I would prefer consistent behavior either way: _either we error when 3D fields are expected or not, regardless of grid type._ I guess the presence of the extra LLC face dimension throws a wrench into the logic somewhere.

rabernat commented on xmitgcm/test/test_mds_store.py:
```diff
@@ -195,6 +196,44 @@ def test_open_dataset_no_meta(all_mds_datadirs):
+
+        # currently raises ONLY with LLC-type datasets, but not regular grids
+        # WHY?!?
+        ds = xmitgcm.open_mdsdataset(dirname, prefix=['T'], **kwargs)
```
My question for the group, including @ifenty, is, what sort of API do we want here? Do we want it to "just work" when these 2D diags are encountered, or do we want to raise an error by default? In this case, we would have a new keyword argument like `allow_2d_diags=True` which would turn on the new feature of correctly parsing 2D slices of 3D diags.

timothyas replied:
My opinion is to require the keyword `allow_2d_diags`. Along with this, we can use the if statements above to print out a useful error message. Something like 

```
#Expect 3 dims, vertical and 2 horizontal 
ndims_expected = 3 
if self.llc:
    # llc geometry adds face dim 
    ndims_expected += 1 
  
if len(dims) == 3 and data.ndim == ndims_expected-1:
    if allow_2d_diags: 
        dims = dims[1:] 
        vname = '%s_2D' % vname 
    else: 
        raise TypeError('Variable %s from file prefix %s has unexpected number of dimensions. \n'
                        'Set allow_2d_diags=True to read 2D diagnostic of a typically 3D variable' % (vname,prefix))
```
Perhaps a different error should be used, but this is my general idea. 

I'd also be happy to turn the notebook I've been using for testing into documentation.

timothyas commented:
Thanks @rabernat for getting me started on the tests!

timothyas commented:
> I just added this test:
> 
> https://github.com/xgcm/xmitgcm/blob/2fd7ad7a8f8c1077e314aa7a0ee54f1d9dbe6e37/xmitgcm/test/test_mds_store.py#L199-L234
> 
> This is not totally complete, but I think it should help @timothyas move forward.
> The strange thing is that it passes just fine already with standard grids but doesn't work with LLC grids. I would prefer consistent behavior either way: _either we error when 3D fields are expected or not, regardless of grid type._ I guess the presence of the extra LLC face dimension throws a wrench into the logic somewhere.

I think I found why it fails for llc but not others. This snippet lived just a few lines above our little hack in `load_from_prefix`: 
https://github.com/xgcm/xmitgcm/blob/2fd7ad7a8f8c1077e314aa7a0ee54f1d9dbe6e37/xmitgcm/mds_store.py#L670-L679

I think this was to catch output from e.g. barotropic simulations, where we actually have 2 dims, even for grid variables like hFac. I have made a number of changes, I went ahead and implemented what I mentioned earlier: added a keyword `allow_2d_diags`. I then removed the above lines, but then had to add the `allow_2d_diags` to the `expected` dictionaries in the test suite. This should allow us to say "yes we expect 2D diags for the barotropic experiment, but no others".  

However I'm also getting a KeyError: with variable 'T' in get_virtual_variable that I don't quite understand. My hunch is that this has to do with the new test, something to do with the renaming and copying of Eta, but I'm at a loss for that one.

timothyas commented:
PS perhaps that was assuming of me to just go ahead and implement what I suggested, but I got addicted to solving this. Happy to revert changes if everyone disagrees.

rabernat commented:
@timothyas - thanks a lot for working on this!

It's not an option to suddenly require a new keyword argument `allow_2d_diags=True` for barotropic simulations that previously worked fine. This would be considered a "breaking change," because it would cause xmitgcm code that previously worked to suddenly stop working if a user upgrades xmitgcm. Imagine the frustration that would cause for you as a user! These sorts of changes must be avoided at all costs. Instead, we should make whatever changes we make here backwards compatible. If we really need to introduce breaking changes, we need to do so very slowly, over many release cycles, starting first with deprecation warnings. I don't think that is necessary here.

I would back up a little bit and rethink the implementation here.

The *default* behavior (`allow_2d_diags=False`) should just leave things as they were before. `allow_2d_diags=True` should turn on a new code path that overcomes the specific problem raised in #140.

> My hunch is that this has to do with the new test, something to do with the renaming and copying of Eta, but I'm at a loss for that one.

Hmm, there could be a problem with my test then. Unfortunately I will not have time to work on this today.

@ifenty - where do we stand on the ECCO summer school. Is there a specific date by which you need this fixed? Have we already missed it? What happens if we fail to implement this new feature by the target date? What work-arounds might exist?

timothyas commented:
Hey @rabernat, thanks for this thorough explanation. This totally makes sense - of course we don't want to break old code! And it makes sense that you would want these tests to break in order to catch breaking changes like this. I'm learning ... slowly... :) 

So I'm now thinking about what was already in the function 
https://github.com/xgcm/xmitgcm/blob/2fd7ad7a8f8c1077e314aa7a0ee54f1d9dbe6e37/xmitgcm/mds_store.py#L670-L679

and it seems like the default behavior is the equivalent of `allow_2d_diags=True`, but now we're just also trying to catch the LLC case. 

I also don't have a ton of time to work on this right now since I have to prepare some other materials for the school. I'm curious what @ifenty thinks, but I would like to propose the following temporary solution. The students will be using a central distribution of xmitgcm, which is in a community directory on the super computer cloned from git. We could just clone a branch (e.g. from my fork) which has this fix for LLC geometries and roll with that.

ifenty commented:
@timothyas  

if that's the best we can do before Monday then we'll give the students your xmitgcm branch.

rabernat commented:
We can easily release a new xmitgcm version over the week.

But I still don't understand...what is the actual fix? Do we know what it is, or it still has to be solved.

I have a hunch it is just a few lines of code.

rabernat commented:
@timothyas - I made an alternate fix for this in #145.

I really appreciate the effort you put into this. It was definitely not for nothing, as you have learned probably more than you ever wanted to know about the guts of xmitgcm. I think you were on track to reaching the same solution. But you mentioned being pressed for time, so I went ahead with my own solution.

timothyas commented:
Thanks @rabernat! Yeah I was thinking the same thing, but you wrote it much cleaner, and fixed the rest which I wasn’t sure on. I’m happy enough with the knowledge gained.
